### ISSUE TITLE
Sharphook not working in published version of app

### ISSUE THREAD

**Issue by TolikPylypchuk** [TolikPylypchuk/SharpHook#90]
### Discussed in https://github.com/TolikPylypchuk/SharpHook/discussions/89

<div type='discussions-op-text'>

<sup>Originally posted by **BlackCoyote** March  6, 2024</sup>
Hey,

We're trying to build an Avalonia app that uses Sharphook, but for some reason we're getting some unexpected behaviour with Sharphook in the published version of the app.

Everything works as expected in the normal debug and release builds of the app, but once we try the publish output, any call to `EventSimulator.SimulateKeyPress()` will not work. It will throw the following exception for us:
```
System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
   at SharpHook.Native.UioHook.PostEvent(UioHookEvent& e)
   at SharpHook.Providers.UioHookProvider.PostEvent(UioHookEvent& e)
   at SharpHook.EventSimulator.SimulateKeyPress(KeyCode keyCode)
   at A.f.A.A()
   ```

When we tried using the [logging](https://sharphook.tolik.io/v5.0.0/articles/logging.html#using-high-level-types) feature of sharphook, we experience a similar exception:
```
System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
   at SharpHook.Native.UioHook.SetLoggerProc(LoggerProc loggerProc, IntPtr userData)
   at SharpHook.Providers.UioHookProvider.SetLoggerProc(LoggerProc loggerProc, IntPtr userData)
   at SharpHook.Logging.LogSource.RegisterOrGet(ILoggingProvider loggingProvider, LogLevel minLevel)
   at SharpHook.Logging.LogSource.RegisterOrGet(LogLevel minLevel)
   at A.O.a()
   ```
   
   Here are the settings in the publish profile being used:
   ```
  <Configuration>Pro</Configuration>
  <Platform>Any CPU</Platform>
  <PublishDir>publish\windows\pro</PublishDir>
  <PublishProtocol>FileSystem</PublishProtocol>
  <_TargetId>Folder</_TargetId>
  <TargetFramework>net6.0</TargetFramework>
  <RuntimeIdentifier>win-x86</RuntimeIdentifier>
  <SelfContained>true</SelfContained>
  <IncludeNativeLibrariesForSelfExtract>true</IncludeNativeLibrariesForSelfExtract>
```
    
    I'm hoping you might have any idea of what we're doing wrong, or if you had any suggestions we might try.
    
    Thanks!</div>

**Comment by TolikPylypchuk** [comment]
Hi @BlackCoyote! `BadImageFormatException` usually means that the bitness of the application doesn't match the bitness of the native library.

Since your application uses a platform-specific library, you shouldn't use `Any CPU` as the platform. Instead, use `x86` or `x64`, though I'm not sure it will help since you've specified the RID as `win-x86`.

Could you please try publishing an x64 version and see if there's the same problem? If not, then it's a problem specifically with x86. You can then publish x64, look at the size of `uiohook.dll`, and then publish x86, and look at the size of `uiohook.dll` again - if the sizes are the same, then for some reason, the x64 version ends up in the publish folder. Instead, if the same problem persists with x64 as well, then we'll have to think why.

Also, if you could provide a minimal reproducible example, it would be ideal.

**Comment by BlackCoyote** [comment]
Thanks again for the quick reply!

It does indeed work when publishing with x64, and the uiohook.dll seems to be 647kB on both x64 and x86.

I've managed to replicate the exact same issue in a fresh net6.0 console project consisting only of Sharphook and a publish profile. It will work for x64 but not for x86, giving the same exception.

```
using SharpHook;
using SharpHook.Logging;
using SharpHook.Native;

namespace SharphookTest
{
    internal class Program
    {
        static void Main(string[] args)
        {
            var logSource = LogSource.RegisterOrGet(minLevel: LogLevel.Info);
            logSource.MessageLogged += (s, e) => Console.WriteLine(e.LogEntry.FullText);
            new EventSimulator().SimulateKeyPress(KeyCode.VcX);
            Console.ReadLine();
        }
    }
```

The publish profile:

```
<Project>
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>win-x86</Platform>
    <PublishDir>bin\Release\net6.0\publish\win-x86\</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <_TargetId>Folder</_TargetId>
    <TargetFramework>net6.0</TargetFramework>
    <RuntimeIdentifier>win-x86</RuntimeIdentifier>
    <SelfContained>false</SelfContained>
    <PublishSingleFile>false</PublishSingleFile>
    <PublishReadyToRun>false</PublishReadyToRun>
  </PropertyGroup>
</Project>
```

Let me know if there's anything else you'd like me to test!

**Comment by TolikPylypchuk** [comment]
Thanks for the reproducible example, I will look into it in the next couple days

In the meantime, as a workaround, you can manually substitute `uiohook.dll` with the x86 version from the package after publishing (not ideal, but better than nothing I guess).

**Comment by TolikPylypchuk** [comment]
I'm still investigating, but the interesting thing is that if you don't use a publish profile, but instead use .NET CLI, then the x86 version is published with the correct `uiohook.dll` 😄 

But I was able to reproduce the issue when publishing through Visual Studio, so will investigate further.

**Comment by TolikPylypchuk** [comment]
OK, I've found another interesting thing. It looks like it's a bug in SharpHook after all.

SharpHook uses a [targets file](https://github.com/TolikPylypchuk/SharpHook/blob/192f8eac5fa3c5def722c94317e233963f5beaa0/SharpHook/SharpHook.targets) to copy `uiohook.dll` into the output folder. The thing is that this targets file is only needed for .NET Framework since .NET can manage to do that on its own. I've configured it to execute only for .NET Framework projects. Or so I've thought.

Well, it appears that the targets file is used when publishing a .NET project via Visual Studio (but not .NET CLI) and it substitutes the correct `uiohook.dll` with a wrong one. I'll fix that, but no promises on the timeline since the problem of unpacking native dependencies from NuGet packages is really difficult to solve correctly for .NET Framework and apparently nobody really knows how to do that.

There's another workaround you can currently use - add the following property to the publish profile:

```xml
<SharpHookPlatform>x86</SharpHookPlatform>
```

This is one of the properties that ShaprHook uses internally in the targets file to determine which native DLL to unpack. Apparently, it does so in a wrong way. I've tried adding it and now the project is published with the correct `uiohook.dll` with Visual Stuio. But I have to warn you that **this property is internal, and so it's subject to change without notice at any moment**. Use it at your own discretion. Or maybe consider using .NET CLI instead 😄

**Comment by BlackCoyote** [comment]
Awesome work! 

Thanks for giving each issue such time and care. And thank you for sharing the temporary workaround and for your suggestion to look into .NET CLI. We'll be sure to make the most of that.

Hopefully the NuGet dependency unpacking won't give you too much trouble!

**Comment by TolikPylypchuk** [comment]
Hello @BlackCoyote! I have fixed the issue, and the fix will be available in the next release, though I'm not sure yet when exactly it will happen as there is another bug to look into.

**Comment by TolikPylypchuk** [comment]
I've just released version 5.3.2 which fixes this issue - you no longer need to specify `SharpHookPlatform` in the publish profile, it should work without it.
